test/integration/test/deck_test.go

```diff
@@ -379,137 +381,137 @@ func TestRerun(t *testing.T) {
-	// configmap updates on existing pods every minute, which is a long wait.
-	// The proposed fix in the issue was updating the deployment, which imo
-	// should be better handled by just refreshing pods.
-	// So here comes forcing restart of horologium pods.
-	if err := refreshProwPods(kubeClient, context.Background(), "horologium"); err != nil {
-		t.Fatalf("Failed refreshing horologium pods: %v", err)
-	}
-	// Same with deck
-	if err := refreshProwPods(kubeClient, context.Background(), "deck"); err != nil {
-		t.Fatalf("Failed refreshing deck pods: %v", err)
+	redeployJobConfig := func(jobConfig string) {
+		if err := updateJobConfig(ctx, kubeClient, rerunJobConfigFile, []byte(jobConfig)); err != nil {
+			t.Fatalf("Failed update job config: %v", err)
+		}
+
+		// Now we are waiting on Horologium to create the first prow job so that we
+		// can rerun from.
+		// Horologium itself is pretty good at handling the configmap update, but
+		// not kubelet, according to
+		// https://github.com/kubernetes/kubernetes/issues/30189 kubelet syncs
+		// configmap updates on existing pods every minute, which is a long wait.
+		// It's quicker to rollout the affected Deployments.
+		if err := rolloutDeployment(t, ctx, kubeClient, "horologium"); err != nil {
+			t.Fatalf("Failed rolling out Horologium: %v", err)
+		}
+		// Same with deck
+		if err := rolloutDeployment(t, ctx, kubeClient, "deck"); err != nil {
+			t.Fatalf("Failed rolling out Deck: %v", err)
+		}
 	}
 
+	// Deploy the initial template with "foo" as the label.
+	redeployJobConfig(fmt.Sprintf(rerunJobConfigTemplate, jobName, "foo"))
+
 	t.Cleanup(func() {
-		if err := updateJobConfig(context.Background(), kubeClient, rerunJobConfigFile, []byte{}); err != nil {
+		if err := updateJobConfig(ctx, kubeClient, rerunJobConfigFile, []byte{}); err != nil {
 			t.Logf("ERROR CLEANUP: %v", err)
 		}
-		labels, err := labels.Parse("prow.k8s.io/job = " + jobName)
-		if err != nil {
-			t.Logf("Skip cleaning up jobs, as failed parsing label: %v", err)
-			return
+		// Prevent horologium from immediately creating the "missing" ProwJob after the
+		// DeleteAll call further down, because horologium still runs with the last
+		// non-empty configuration (foo=bar).
+		if err := rolloutDeployment(t, ctx, kubeClient, "horologium"); err != nil {
+			t.Logf("Failed rolling out Horologium: %v", err)
 		}
-		if err := kubeClient.DeleteAllOf(context.Background(), &prowjobv1.ProwJob{}, &ctrlruntimeclient.DeleteAllOfOptions{
-			ListOptions: ctrlruntimeclient.ListOptions{LabelSelector: labels},
+		if err := kubeClient.DeleteAllOf(ctx, &prowjobv1.ProwJob{}, &ctrlruntimeclient.DeleteAllOfOptions{
+			ListOptions: ctrlruntimeclient.ListOptions{
+				Namespace:     defaultNamespace,
+				LabelSelector: prowJobSelector,
+			},
 		}); err != nil {
 			t.Logf("ERROR CLEANUP: %v", err)
 		}
 	})
-	ctx := context.Background()
+
 	getLatestJob := func(t *testing.T, jobName string, lastRun *v1.Time) *prowjobv1.ProwJob {
```

Comment:
It seems like we're missing the meaning of what this function was originally written to do - is the issue that the function expected to sort these resources by the `resourceVersion` at which they were created, but when there are interspersed UPDATE calls, the objects' current `resourceVersion` no longer sorts them?

Can we sort by the job ID since we know that's monotonically increasing? Creation timestamp is an awkward choice as it can have ties.

Reply:
Sorry for the long response delay.

Which "job ID" are you referring to exactly? The only numerical ID I can see is the `prow.k8s.io/build-id` and that is not unique per ProwJob. Is `prow.k8s.io/id` not just a random UUID, but monotonically increasing?

I don't see any other good value to use besides the creation timestamp when I want to search by, well, creation order:

```yaml
apiVersion: prow.k8s.io/v1
kind: ProwJob
metadata:
  annotations:
    prow.k8s.io/context: ""
    prow.k8s.io/job: rerun-test-job-3a2c5361172244414edc254fc8d21de5
  creationTimestamp: "2024-12-14T15:54:56Z"
  generation: 3
  labels:
    created-by-prow: "true"
    foo: foo
    prow.k8s.io/build-id: "1867961480935116800"
    prow.k8s.io/context: ""
    prow.k8s.io/id: f48bd245-1335-4ef4-8556-d7b9d232014a
    prow.k8s.io/job: rerun-test-job-3a2c5361172244414edc254fc8d21de5
    prow.k8s.io/type: periodic
  name: f48bd245-1335-4ef4-8556-d7b9d232014a
  namespace: default
  resourceVersion: "9383"
  uid: 56b93b1c-4a53-4ba6-90e6-49ce918115cd
spec:
  agent: kubernetes
  cluster: default
  job: rerun-test-job-3a2c5361172244414edc254fc8d21de5
  namespace: test-pods
  pod_spec:
    containers:
    - args:
      - Hello World!
      command:
      - echo
      image: localhost:5001/alpine
      name: ""
      resources: {}
  prowjob_defaults:
    tenant_id: GlobalDefaultID
  report: true
  type: periodic
status:
  build_id: "1867961480935116800"
  completionTime: "2024-12-14T15:54:58Z"
  description: Job succeeded.
  pendingTime: "2024-12-14T15:54:56Z"
  pod_name: f48bd245-1335-4ef4-8556-d7b9d232014a
  startTime: "2024-12-14T15:54:56Z"
  state: success
```